src/AI/JakeMan.java

```diff
@@ -0,0 +1,770 @@
+            final Unit defender = results.defender.unit;
+            int loss   = Math.min(unit    .getHP(), (int)results.attacker.getPreciseHPDamage());
+            int damage = Math.min(defender.getHP(), (int)results.defender.getPreciseHPDamage());
+
+            // Convert to abstract value
+            int extraLoss = 0;
+            if( loss >= 1 && unit.getHP() == UnitModel.MAXIMUM_HP )
+              extraLoss += STAY_UNHURT_BIAS;
+            if( loss >= unit.getHP() )
+              extraLoss += STAY_ALIVE_BIAS;
+            loss *= unit.getCost();
+            loss += extraLoss;
+            damage *= defender.getCost();
+            if( isThreatenedBy(unit.model, defender.model) )
+              damage *= FIRSTSTRIKE_ON_THREAT_WEIGHT;
+            // Value damage to hurt units less
+            damage *= defender.getHPFactor();
+
+            return (double)(damage - loss);
+          }, (terrain, params) -> 0.01); // Attack terrain, but don't prioritize it over units
+
+      if( damageValue > bestDamage )
+      {
+        bestDamage = damageValue;
+        bestAttack = attack;
+      }
+    }
+    return bestAttack;
+  }
+
+  private static boolean isThreatenedBy(UnitModel um, UnitModel threat)
+  {
+    int threshhold = um.hasDirectFireWeapon() ? DIRECT_THREAT_THRESHHOLD : INDIRECT_THREAT_THRESHHOLD;
+    boolean isThreat = false;
+    for( WeaponModel wm : threat.weapons )
+      isThreat |= threshhold <= wm.getDamage(um);
+    return isThreat;
+  }
+
+  private static boolean isWeakTo(UnitModel um, UnitModel threat)
+  {
+    boolean isWeak = isThreatenedBy(um, threat);
+    isWeak &= !isThreatenedBy(threat, um);
+    return isWeak;
+  }
+
+  private boolean isDudeFree(GameMap gameMap, Unit unit, XYCoord xyc, boolean amAttacking)
+  {
+    HashMap<UnitModel, Double> threatCounts = new HashMap<>();
+    for( UnitModel threat : unitMapEnemy.keySet() )
+    {
+      if( !isThreatenedBy(unit.model, threat) )
+        continue;
+      if( unitMapEnemy.get(threat).containsKey(xyc) )
+        threatCounts.put(threat, unitMapEnemy.get(threat).get(xyc));
+    }
+    if( threatCounts.size() < 1 )
+      return true;
+
+    ArrayList<XYCoord> counterCoords = Utils.findLocationsInRange(gameMap, xyc, 1);
+    counterCoords.remove(xyc);
+    for( UnitModel threat : threatCounts.keySet().toArray(new UnitModel[0]) )
+      for( UnitModel counter : unitMapFriendly.keySet() )
+      {
+        if( !isThreatenedBy(threat, counter) )
+          continue;
+        double counterPowerTotal = 0;
+        final boolean counterIsMeAndIAmPeaceful = !amAttacking && (counter == unit.model);
+        for( XYCoord coord : counterCoords )
+        {
+          double counterPower = 0;
+          if( unitMapFriendly.get(counter).containsKey(coord) )
+            counterPower = unitMapFriendly.get(counter).get(coord);
+          if( counterIsMeAndIAmPeaceful )
+            counterPower -= PEACEFUL_SELF_THREAT_RATIO * unit.getHPFactor();
```

Comment:
Why create a global constant for this? Seems like any other value wouldn't make much sense here.

Reply:
Because I was tweaking it and seeing what it did to the behavior.
I think I ended up finding the issue I was trying to solve somewhere else.